src/Chainweb/BlockHeaderDB.hs

```diff
@@ -194,6 +203,156 @@ instance HasChainwebVersion BlockHeaderDb where
+        -- is not necessarily included in the overall consensus of the chainweb.
+        -- In particular the the higest ranking entry is not necessarily the
+        -- entry of largest POW weight.
+        --
+        -- Usually this number would be defined in terms of the chainweb
+        -- diameter and/or the difficulty adjustment window.
+    -> IO Int
+pruneForks logger cdb callback limit = do
+
+    -- find all roots at \(maxEntry - limit\)
+    --
+    m <- maxRank cdb
+    let rootHeight = m - min m limit
+
+    roots <- keys cdb Nothing Nothing
+        (Just $ MinRank $ Min rootHeight)
+        (Just $ MaxRank $ Max rootHeight)
+        streamToHashSet_
+
+    -- Bloom filter for marking block headers
+    --
+    let s = max (int rootHeight + 10 * HS.size roots) (int rootHeight * 2)
+    let (size, hashNum) = BF.suggestSizing s 0.0001
+    marked <- stToIO $ BF.new (BF.cheapHashes hashNum) size
+
+    -- Bloom filter for marking block payload hashes
+    --
+    -- Payloads can be shared between different blocks. Thus, to be on the safe
+    -- side, we are not deleting any marked payload.
+    --
+    -- Note that we could use cuckoo hashes or counting bloom filters to do
+    -- reference counting and collect more unreferenced payloads.
+    --
+    -- TODO: would it be better to use a single shared filter?
+    --
+    let (psize, phashNum) = BF.suggestSizing s 0.0001
+    markedPayloads <- stToIO $ BF.new (BF.cheapHashes phashNum) psize
+
+    -- Iterate backwards and mark all predecessors of the roots
+    --
+    void $ branchEntries cdb Nothing Nothing Nothing Nothing mempty (HS.map UpperBound roots)
+        $ S.mapM_ $ \h -> do
+            stToIO $ do
+                BF.insert marked $ runPut $ encodeBlockHash $ _blockHash h
+                BF.insert markedPayloads
+                    $ runPut $ encodeBlockPayloadHash $ _blockPayloadHash h
+
+    -- Iterate forward and delete all non-marked block headers that are not a
+    -- predecessor of a block header that is kept.
+    --
+    entries cdb Nothing Nothing Nothing (Just $ MaxRank $ Max rootHeight)
+        $ S.foldM_ (go marked markedPayloads) (return (mempty, 0)) (return . snd)
+
+  where
+    logg = logFunctionText (setComponent "pact-tx-replay" logger)
+
+    -- The falsePositiveSet collects all deleted nodes that are known to be
+    -- false positives. We know that a node is false positive when it is in the
+    -- filter but any of its ancestors is either not in the fitler or in the
+    -- falsePositiveSet. Note that this doesn't capture all false positives, but
+    -- only those that are not connected to the main chain.
+    --
+    -- Nodes that are known to be false positives are safe to remove, if also
+    -- all of its successors are removed.
+    --
+    -- Nodes that are know to be false postives must be removed, if any of their
+    -- predecessors got removed.
+    --
+    go marked markedPayloads (falsePositiveSet, i) h = do
```

Comment:
You're folding `go` so put bangs on your accumulators here (`falsePositiveSet` and `i`)